dockerfiles/theia/src/main.js

```diff
@@ -11,46 +11,159 @@
+
+/** Returns standard (default) Theia dependencies */
+function getDefaultTheiaPlugins() {
+    const defaultTheiaPackageJson = require(`${defaultTheiaRoot}/package.json`);
+    return defaultTheiaPackageJson['dependencies'];
+}
+
+/** Returns current (after previous launch) Theia plugins or empty object on first start */
+function getTheiaPlugins() {
+    if (!fs.existsSync(`${theiaRoot}/package.json`)) {
+        return {};
+    }
+    const theiaPackageJson = require(`${theiaRoot}/package.json`);
+    return theiaPackageJson['dependencies'];
+}
+
+/** Returns extra Theia plugins which is set via THEIA_PLUGINS env variable */
+function getExtraTheiaPlugins() {
+    let plugins = {};
+    const currentPluginsString = process.env.THEIA_PLUGINS;
+    if (currentPluginsString) {
+        for (let plugin of currentPluginsString.split(',')) {
+            plugin = plugin.trim();
+            if (plugin === '') {
                 continue;
             }
-            if (!dep.hasOwnProperty(d)) {
-                dep[d] = "latest";
+            const colonPos = plugin.indexOf(':');
+            if (colonPos !== -1) {
+                const pluginName = plugin.substring(0, colonPos).trim();
+                const pluginVersion = plugin.substring(colonPos + 1).trim();
+                if (pluginVersion.startsWith('git://') || pluginVersion.startsWith('https://')) {
+                    addPluginFromGitRepository(plugins, pluginName, pluginVersion);
+                } else {
+                    plugins[pluginName] = pluginVersion;
+                }
+            } else {
+                plugins[plugin] = 'latest';
             }
         }
-        fs.writeFileSync(theiaPath, JSON.stringify(theiaConfig));
-        handlePromise(callYarn().then(callBuild).then(callRun));
-    } else {
-        cp.execSync(`rsync -rv ${defaultTheiaPath}/ ${theiaRoot} --exclude 'node_modules' --exclude 'yarn.lock'`);
-        handlePromise(callRun());
+    }
+    return plugins;
+}
+
+/**
+ * Clones git repository and adds the plugin as local.
+ * This is done becouse Theia requires npm package.json (not lerna) which is located, in most cases, in subdirectory of repository.
+ */
+function addPluginFromGitRepository(plugins, pluginName, gitRepository) {
+    const pluginPath = gitPluginsRoot + '/' + pluginName + '/';
+    try {
+        cp.execSync(`git clone --depth=1 --quiet ${gitRepository} ${pluginPath}`);
+    } catch (error) {
+        // failed to clone repository
+        try {
+            cp.execSync(`git clone --depth=1 --quiet ${gitRepository} ${pluginPath}`);
+        } catch (error) {
+            // failed again, skip plugin
+            console.error('Failed to get plugin: ' + pluginName + '. Skipping... Is the url specified properly?');
+            return;
+        }
     }
 
+    const rootPackageJson = require(pluginPath + 'package.json');
+    if (rootPackageJson['name'] === pluginName && !fs.existsSync(pluginPath + 'lerna.json')) {
```

Comment:
Seems in this case we need check if node_modules and lib exists and build plugin.  We can split  build logic to separated method and reuse here. @mmorhun  What do you think?

Reply:
Well, we do not know how to build this plugin in general case. And most of the plagins (at least generated with Theai plugin generator, which is majority) use lerna (i.e. else block will be executed). I use this just to be able to include kinda 'ready' plugin (as you know, will work with the plugin's subdirectory). This can be used to speed up Theia run.